Feat :support create service

Mcduller opened the pull request:
### Description of your changes

<!--

Briefly describe what this pull request does. We love pull requests that resolve an open KubeVela issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->

In order to support future webhook sources, support for creating services, and some code refactoring

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [x] Add related tests.
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->


### Special notes for your reviewer

<!--

Be sure to direct your reviewers'
attention to anything that needs special consideration.

-->

Mcduller committed:
Test:support create service

Signed-off-by: Mcduller <1596582524@qq.com>

Mcduller committed:
Test:support create service

Signed-off-by: Mcduller <1596582524@qq.com>

FogDong commented on controllers/triggerservice/triggerservice_controller.go:
```diff
@@ -91,25 +92,36 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
+		return err
+	}
+
+	if err := r.createService(ctx, ts, v); err != nil {
```
Please check `createService` first, if the value is true, then create the service.

Mcduller replied:
you mean check before invoking the function? or we can check the value in the function ?

FogDong commented on config/definition/default.yaml:
```diff
@@ -147,4 +181,13 @@ spec:
         }]
+        createService: *true | bool
+        service?: {
+          ports:[...{
```
```suggestion
          ports: *[{
                  name: "default"
                  port: 80
                  targetPort: 80
                }] | [...{
                   name: string
                   port: int
                   targetPort: int
                }]
```

FogDong commented on config/definition/default.yaml:
```diff
@@ -147,4 +181,13 @@ spec:
+            port: int
+            targetPort: int|string
+          }]
+          clusterIp?: string
```
```suggestion
  type: *"ClusterIP" | "NodePort" | "LoadBalancer" | "ExternalName"
```

FogDong commented on config/definition/default.yaml:
```diff
@@ -147,4 +181,13 @@ spec:
           hostNames?: [...string]
         }]
+        createService: *true | bool
+        service?: {
```
```suggestion
        service: {
```

FogDong commented on config/definition/default.yaml:
```diff
@@ -109,6 +109,40 @@ spec:
+                  }
+                }]
+              if parameter.service.clusterIp != _|_ {
+                  clusterIp: parameter.service.clusterIp
```
Why users need to specify the clusterIP?

Mcduller replied:
> Why users need to specify the clusterIP?
my fault, i will remove it.

Mcduller committed:
Update config/definition/default.yaml

Co-authored-by: Tianxin Dong <wuwuglu19@gmail.com>

Mcduller committed:
Update config/definition/default.yaml

Co-authored-by: Tianxin Dong <wuwuglu19@gmail.com>

Mcduller committed:
Update config/definition/default.yaml

Co-authored-by: Tianxin Dong <wuwuglu19@gmail.com>

Mcduller committed:
Update config/definition/default.yaml

Co-authored-by: Tianxin Dong <wuwuglu19@gmail.com>

Mcduller committed:
Update config/definition/default.yaml

Co-authored-by: Tianxin Dong <wuwuglu19@gmail.com>

Mcduller committed:
Update config/definition/default.yaml
add unit test

Co-authored-by: Tianxin Dong <wuwuglu19@gmail.com>
Signed-off-by: Mcduller <1596582524@qq.com>

Mcduller committed:
Update config/definition/default.yaml
add unit test

Co-authored-by: Tianxin Dong <wuwuglu19@gmail.com>
Signed-off-by: Mcduller <1596582524@qq.com>

Mcduller committed:
Merge remote-tracking branch 'origin/create_svc' into create_svc

# Conflicts:
#	config/definition/default.yaml

Mcduller committed:
fix unit test

Signed-off-by: Mcduller <1596582524@qq.com>

semmet95 committed:
Fix: example manager and instructions (#48)

* Feat: upgrade k8s.io dependency to 0.26 (#39)

* Feat: upgrade k8s.io dependency to 0.26

Signed-off-by: Yin Da <yd219913@alibaba-inc.com>

* Fix: golangci-lint action failure

Signed-off-by: Yin Da <yd219913@alibaba-inc.com>

---------

Signed-off-by: Yin Da <yd219913@alibaba-inc.com>
Signed-off-by: Amit Singh <singhamitch@outlook.com>

* Feat: add cluster info to action context (#43)

Feat: add cluster info in action context

Signed-off-by: yangsoon <songyang.song@alibaba-inc.com>
Co-authored-by: yangsoon <songyang.song@alibaba-inc.com>
Signed-off-by: Amit Singh <singhamitch@outlook.com>

* Feat: add cronjob source (#42)

* Refactor: allow the use of singleton for one Source, while also allowing non-singleton Sources, making it more extensible

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* Feat: add cronjob source

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* Chore: update cron lib

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* Feat: do not exit if config contains invalid sources

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* Chore: add tests and fix linter

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* Chore: go mod tidy

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* Docs: add cronjob example

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* fix comments

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* Chore: force lint to use the exact version

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* remove unnecessary dep

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* fix comments

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* use codecov token

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

* update lint script

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

---------

Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>
Signed-off-by: Amit Singh <singhamitch@outlook.com>

* Feat: support leader election for kube-trigger (#45)

Signed-off-by: yangsoon <songyang.song@alibaba-inc.com>
Co-authored-by: yangsoon <songyang.song@alibaba-inc.com>
Signed-off-by: Amit Singh <singhamitch@outlook.com>

* adds instruction to enable kube-trigger addon

Signed-off-by: Amit Singh <singhamitch@outlook.com>

* updates sample.yaml path

Signed-off-by: Amit Singh <singhamitch@outlook.com>

* fix: updates instructions to apply config files in order

Signed-off-by: Amit Singh <singhamitch@outlook.com>

* fix: adds the missing api group to the first rule

Signed-off-by: Amit Singh <singhamitch@outlook.com>

* Revert "Feat: support leader election for kube-trigger (#45)"

This reverts commit 4351008607f9f311cb43f1132292cfa42746d57a.

Removing changes unrelated to this branch

Signed-off-by: Amit Singh <singhamitch@outlook.com>

* Revert "Feat: add cronjob source (#42)"

This reverts commit 46f5975b64e9c3e412e7d3e5564b08d111814ce5.

Removes changes unrelated to the branch

Signed-off-by: Amit Singh <singhamitch@outlook.com>

* Revert "Feat: add cluster info to action context (#43)"

This reverts commit e214be5be37d60222170eb416c3234135ef994bf.

Removes changes unrelated to the branch

Signed-off-by: Amit Singh <singhamitch@outlook.com>

* Revert "Feat: upgrade k8s.io dependency to 0.26 (#39)"

This reverts commit a22e31611efdf53ad8029e254dfea48ba0cf32ac.

Removes changes unrelated to the branch

Signed-off-by: Amit Singh <singhamitch@outlook.com>

---------

Signed-off-by: Yin Da <yd219913@alibaba-inc.com>
Signed-off-by: Amit Singh <singhamitch@outlook.com>
Signed-off-by: yangsoon <songyang.song@alibaba-inc.com>
Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>
Co-authored-by: Somefive <Somefive@foxmail.com>
Co-authored-by: yangs <yangsoonlx@gmail.com>
Co-authored-by: yangsoon <songyang.song@alibaba-inc.com>
Co-authored-by: Charlie Chiang <charlie_c_0129@outlook.com>

Mcduller committed:
fix unit test

Signed-off-by: Mcduller <1596582524@qq.com>

Mcduller committed:
Merge remote-tracking branch 'origin/create_svc' into create_svc

Mcduller committed:
Merge remote-tracking branch 'origin/create_svc' into create_svc

Mcduller committed:
fix unit test

Signed-off-by: Mcduller <1596582524@qq.com>

Mcduller committed:
fix unit test

Signed-off-by: Mcduller <1596582524@qq.com>

Mcduller commented:
@charlie0129 @FogDong please review this pr  :)

Mcduller committed:
Merge remote-tracking branch 'origin/create_svc' into create_svc

# Conflicts:
#	controllers/triggerservice/suite_test.go
#	controllers/utils/test_util.go

Mcduller committed:
fix unit test

Signed-off-by: Mcduller <1596582524@qq.com>

Mcduller committed:
Merge remote-tracking branch 'origin/create_svc' into create_svc

# Conflicts:
#	controllers/utils/test_util.go

Mcduller committed:
add unit test restart pod

Signed-off-by: Mcduller <1596582524@qq.com>

Mcduller committed:
Merge remote-tracking branch 'origin/create_svc' into create_svc

FogDong commented on controllers/triggerservice/triggerservice_controller.go:
```diff
@@ -183,6 +208,44 @@ func (r *Reconciler) createRoles(ctx context.Context, ts *standardv1alpha1.Trigg
+		return err
+	}
+
+	if !needCreateService {
```
If the `needCreateService` is false, then we'll try to delete the existing service? I think it is a dangerous behavior and we shouldn't clean the service for user automatically.

Mcduller replied:
should we add another flag to clean the service when user want to clean service？

FogDong commented on config/manager/role.yaml:
```diff
@@ -5,7 +5,9 @@ metadata:
-- resources:
+- apiGroups:
+  - ""
+  resources:
```
Is this change caused by our scripts? /cc @charlie0129
